### PR TITLE
sandbox/cgroup: improve lookup for memory controller support with v2

### PR DESCRIPTION
**Pull request by bboozzoo**
With cgroup v2, we should not be looking at /proc/cgroups looking for memory controller, but rather at the list of availble controllers at the root of the hierarchy.

Thanks for helping us make a better snapd!
Have you signed the [license agreement](https://www.ubuntu.com/legal/contributors) and read the [contribution guide](https://github.com/snapcore/snapd/blob/master/CONTRIBUTING.md)?

Related: [SNAPDENG-34247](https://warthogs.atlassian.net/browse/SNAPDENG-34247)

[SNAPDENG-34247]: https://warthogs.atlassian.net/browse/SNAPDENG-34247?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: sandbox/cgroup/memory.go [hunk 1]
```diff
@@ -51,20 +75,38 @@ func CheckMemoryCgroup() error {
...
+	if !scanner.Scan() {
+		if err := scanner.Err(); err != nil {
+			return false, err
+		}
+		return false, nil
+	}
+
+	line := scanner.Text()
+	controllers := strings.Fields(line)
+	return strutil.ListContains(controllers, "memory"), nil
```
Comment (andrewphelpsj) [inline sandbox/cgroup/memory.go:111]: Maybe something like this? Doesn't matter too much though.
```
scanner := bufio.NewScanner(f)
scanner.Split(bufio.ScanWords)
                                      
for scanner.Scan() {
	if scanner.Text() == "memory" {
		return true, nil
	}
}
                                      
if err := scanner.Err(); err != nil {
	return false, err
}
                                      
return false, nil
```

**Review by andrewphelpsj — APPROVED** [review]
Thanks!

File: sandbox/cgroup/memory.go (hunk 1 above)
Comment (bboozzoo) [reply sandbox/cgroup/memory.go:111]: good idea, applied

File: sandbox/cgroup/memory.go [hunk 2]
```diff
@@ -23,10 +23,21 @@ import (
 	"bufio"
 	"fmt"
 	"os"
+	"path/filepath"
 	"strings"
 )
 
-var cgroupsFilePath = "/proc/cgroups"
+var (
+
```
Comment (zyga) [inline sandbox/cgroup/memory.go:31]: Nitpick, newline.

File: sandbox/cgroup/memory.go [hunk 3]
```diff
@@ -38,9 +49,30 @@ var cgroupsFilePath = "/proc/cgroups"
...
+	if err != nil {
+		return err
+	}
+
+	if supp {
+		return nil
+	}
+
+	// no errors so far but the only path here is the cgroups file without the memory line
+	return fmt.Errorf("cgroup memory controller is disabled on this system")
```
Comment (zyga) [inline sandbox/cgroup/memory.go:69]: This should probably be a variable using `errors.New`

File: sandbox/cgroup/memory.go [hunk 4]
```diff
@@ -38,9 +49,30 @@ var cgroupsFilePath = "/proc/cgroups"
...
+}
+
+func checkV1CgroupMemoryController() (bool, error) {
+	cgroupsFile, err := os.Open(filepath.Join(rootPath, cgroupV1ControllersPath))
 	if err != nil {
-		return fmt.Errorf("cannot open cgroups file: %v", err)
+		return false, fmt.Errorf("cannot open cgroups file: %v", err)
```
Comment (zyga) [inline sandbox/cgroup/memory.go:75]: We can wrap the error:
```suggestion
		return false, fmt.Errorf("cannot open cgroups file: %w", err)
```

File: sandbox/cgroup/memory.go [hunk 5]
```diff
@@ -51,20 +83,40 @@ func CheckMemoryCgroup() error {
...
-				return fmt.Errorf("memory cgroup is disabled on this system")
-			}
-			return nil
+			return isMemoryEnabled, nil
 		}
 	}
 
 	if err := scanner.Err(); err != nil {
-		return fmt.Errorf("cannot read %s contents: %v", cgroupsFilePath, err)
+		return false, fmt.Errorf("cannot read %s contents: %v", cgroupV1ControllersPath, err)
```
Comment (zyga) [inline sandbox/cgroup/memory.go:94]: Ditto

```suggestion
		return false, fmt.Errorf("cannot read %s contents: %w", cgroupV1ControllersPath, err)
```

**Review by zyga — APPROVED** [review]
LGTM with some comments inline.

**Review by Meulengracht — APPROVED** [review]

**Pull request merged**